homeassistant/components/irc.py

```diff
@@ -0,0 +1,303 @@
+# cannot be added after the object has been constructed (like in __init__).
+# It's an implementation detail in a library that irc3 uses, so this is a
+# work-around to make it work with Home Assistant.
+def make_plugin():
+    """Create plugin to intercept messages from IRC server."""
+    import irc3
+
+    # pylint: disable=redefined-outer-name
+    @irc3.plugin
+    class HAPlugin(object):
+        """Plugin to IRC3 that handles IRC commands."""
+
+        # Matches topics when re-connecting to active ZNC session
+        TOPIC_BNC_REPLAY = irc3.rfc.raw.new(
+            'TOPIC_BNC_REPLAY',
+            (r'^(@(?P<tags>\S+) )?:(?P<mask>\S+) 332 '
+             r'(?P<nick>\S+) (?P<channel>\S+) :(?P<data>.+)'))
+
+        # All users currently in a channel (matches NAMES command)
+        NAME_LIST = irc3.rfc.raw.new(
+            'NAME_LIST',
+            (r'^(@(?P<tags>\S+) )?:(?P<mask>\S+) 353 '
+             r'(?P<nick>\S+) . (?P<channel>\S+) :(?P<users>.+)'))
+
+        def __init__(self, context):
+            """Initialize a new HAPlugin instance."""
+            self.context = context
+            self.hass = context.config.get('hass')
+            self.config = context.config.get('hass_config')
+            self.network = self.config.get(CONF_NETWORK)
+            self._channels = {}
+
+        def server_ready(self):
+            """Triggered after the server sent the MOTD."""
+            _LOGGER.info('Connected to IRC network %s',
+                         self.config.get(CONF_NETWORK))
+
+        @staticmethod
+        def connection_lost():
+            """Triggered when connection is lost."""
+            _LOGGER.error('Failed to connect to IRC server')
+
+        def get_channel(self, channel):
+            """Return internal channel object."""
+            if channel not in self._channels:
+                self._channels[channel] = IRCChannel(channel)
+            return self._channels[channel]
+
+        @irc3.event(irc3.rfc.PRIVMSG)
+        def on_privmsg(self, mask=None, target=None, data=None, **kw):
+            """Handle PRIVMSG commands."""
+            # Do not trigger updates for znc playback
+            self._znc_toggle(mask, target, data)
+
+            nick = mask.split('!')[0]
+            self._update_channel(target, last_speaker=nick, last_message=data)
+
+            # Check whitelist before firing an event
+            allowed = self.config.get(CONF_WHITELIST)
+            if not any(map(lambda pattern: re.match(pattern, nick), allowed)):
+                return
+
+            # Check if we must be addressed by user
+            only_when_addressed = self.config.get(CONF_ONLY_WHEN_ADDRESSED)
+            is_addressed = data.startswith('{0}:'.format(self.context.nick))
+            if only_when_addressed and not is_addressed:
+                return
+
+            channel = self.get_channel(target)
+            self.hass.bus.fire(EVENT_PRIVMSG, {
```

Comment:
In fact, `bus.fire` should not even work when called from inside the event loop.